server/src/main/java/io/crate/expression/scalar/HasSchemaPrivilegeFunction.java

```diff
@@ -0,0 +1,285 @@
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+import io.crate.user.Privilege;
+import io.crate.user.User;
+import io.crate.user.UserLookup;
+import org.elasticsearch.common.TriFunction;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+public class HasSchemaPrivilegeFunction extends Scalar<Boolean, Object> {
+
+    public static final String NAME = "has_schema_privilege";
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final UserLookup userLookup;
+
+    @Nullable
+    private BiFunction<UserLookup, Object, User> getUser;
+
+    private TriFunction<User, Object, String, Boolean> checkPrivilege;
+
+    private static final BiFunction<UserLookup, Object, User> userByName = (userLookup, userName) -> {
+        var user = userLookup.findUser((String) userName);
+        if (user == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "User %s does not exist", userName));
+        }
+        return user;
+    };
+
+    private static final BiFunction<UserLookup, Object, User> userByOid = (userLookup, userOid) -> {
+        var user = userLookup.findUser((Integer) userOid);
+        if (user == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "User with OID %d does not exist", userOid));
+        }
+        return user;
+    };
+
+    private static final TriFunction<User, Object, String, Boolean> checkPrivilegeBySchemaName = (user, schemaName, privilege) -> {
+        var privilegesToCheck = privilegesToCheck(privilege);
+        boolean result = false;
+        for (Privilege.Type type: privilegesToCheck) {
+            // Last argument is null as it's not used for Privilege.Clazz.SCHEMA
+            result |= user.hasPrivilege(type, Privilege.Clazz.SCHEMA, (String) schemaName, null);
+        }
+        return result;
+    };
+
+    private static final TriFunction<User, Object, String, Boolean> checkPrivilegeBySchemaOid = (user, schemaOid, privilege) -> {
+        var privilegesToCheck = privilegesToCheck(privilege);
+        boolean result = false;
+        for (Privilege.Type type: privilegesToCheck) {
+            result |= user.hasSchemaPrivilege(type, (Integer) schemaOid); // Returns true if has any privilege out of 2 possible
+        }
+        return result;
+    };
+
+    /**
+     * @param privilege is a comma separated list.
+     * Valid privileges are 'CREATE' and 'USAGE' which map to DDL and DQL correspondingly.
```

Comment:
Used to return "true" for `USAGE` 
as I misunderstood quote from https://www.postgresql.org/docs/current/ddl-priv.html
> USAGE
For schemas, allows access to objects contained in the schema (assuming that the objects' own privilege requirements are also met). Essentially this allows the grantee to “look up” objects within the schema. Without this permission, it is still possible to see the object names, e.g., by querying system catalogs

but as shown in https://www.db-fiddle.com/f/5YRUUseBEQuLDfZf4GKtBc/5, select without granting returns false, so always returning true is clearly incorrect.

changed to USAGE -> DQL in https://github.com/crate/crate/pull/12553/commits/a57b1a84b6c32a414e6246b49e35910268f72745